### PR TITLE
[JENKINS-48012] Require a X-Hub-Signature header when receiving a hook payload and if…

### PR DESCRIPTION
**Pull request by silbernm**
… a secret is configured

As reported in JENKINS-48012, an attacker can currently circumvent the signature check of the GitHub plugin simply by omitting the X-Hub-Signature header.

---
This change is [Reviewable](https://reviewable.io/reviews/jenkinsci/github-plugin/188)

### CODE REVIEW

File: src/test/java/org/jenkinsci/plugins/github/test/HookSecretHelper.java [hunk 1]
```diff
@@ -62,4 +62,20 @@ public void run() {
     public static void storeSecret(final String secretText) {
         storeSecretIn(Jenkins.getInstance().getDescriptorByType(GitHubPluginConfig.class), secretText);
     }
+
+    /**
+     * Unsets the current hook secret.
+     *
+     * @param config where to remove
+     */
+    public static void removeSecretIn(GitHubPluginConfig config) {
```
Comment (daniel-beck) [inline src/test/java/org/jenkinsci/plugins/github/test/HookSecretHelper.java]: `@Restricted` perhaps?

Comment (oleg-nenashev) [reply src/test/java/org/jenkinsci/plugins/github/test/HookSecretHelper.java]: It's in the Test scope

**Comment by daniel-beck** [comment]
Seems to be introduced by #134:

> also don't bother signature validation if no header from github with signature

Versions from v1.21.0 appear to be affected.

**Comment by daniel-beck** [comment]
@oleg-nenashev 

> Currently it says "Checks that an incoming request has a valid signature, if there is specified a signature in the config." which is exactly what the method does.

It only checks the signature validity if it's present, but if there's no signature (which is necessary) it won't fail. Or do I misunderstand this?

**Comment by silbernm** [comment]
@daniel-beck 
@oleg-nenashev 
I think we are on the same page, but the documentation is misleading. There is never a "signature" in the config, but a "hook secret". I understood the Javadoc comment as "if there is specified a secret in the config" (because that is the only thing that makes sense IMHO) and coded my fix according to that interpretation. I have tried coming up with a better wording and I have moved the line that reads the signature header further down, such that it is clear that the header is validated if and only if a hook secret is specified in the GitHub plugin config.

Regarding GitHub Enterprise on-premise: That is exactly the setup that we are using at my workplace. When not configuring a shared secret in the GitHub plugin configuration, the fixed method will continue *not* to verify the signature, even if one was provided. So we should be fine on that use-case.

**Comment by daniel-beck** [comment]
> `[ERROR] src/main/java/org/jenkinsci/plugins/github/webhook/RequirePostWithGHHookPayload.java[135] (sizes) LineLength: Line is longer than 120 characters (found 124).`

**Pull request merged**